modules/replicator/src/main/java/org/apache/ignite/internal/replicator/Replica.java

```diff
@@ -177,35 +166,42 @@ public CompletableFuture<? extends NetworkMessage> processPlacementDriverMessage
     public CompletableFuture<LeaseGrantedMessageResponse> processLeaseGrantedMessage(LeaseGrantedMessage msg) {
         LOG.info("Received LeaseGrantedMessage for replica belonging to group=" + groupId() + ", force=" + msg.force());
 
-        return leaderFuture().thenCompose(leader -> {
-            HybridTimestamp leaseExpirationTime = this.leaseExpirationTime;
-
-            if (leaseExpirationTime != null) {
-                assert msg.leaseExpirationTime().after(leaseExpirationTime) : "Invalid lease expiration time in message, msg=" + msg;
-            }
+        return readyMajority.thenCompose(unused -> raftClient.readLeaderMetadata()).thenCompose(leaderMetadata -> {
+            assert leaseExpirationTime == null || msg.leaseExpirationTime().after(leaseExpirationTime) :
+                    "Invalid lease expiration time in message [leaseExpirationTime=" + leaseExpirationTime
+                            + ", msgLeaseExpirationTime=" + msg.leaseExpirationTime() + ']';
 
             if (msg.force()) {
                 // Replica must wait till storage index reaches the current leader's index to make sure that all updates made on the
                 // group leader are received.
-
-                return waitForActualState(msg.leaseExpirationTime().getPhysical())
+                return waitForActualState(leaderMetadata.getIndex())
                         .thenCompose(v -> {
+                            if (msg.leaseExpirationTime().getPhysical() < currentTimeMillis()) {
```

Comment:
Is this the correct way to compare hybrid timestamp? I was expecting `HybridTimestamp#compareTo` with `clock.now`